datafusion/core/src/datasource/file_format/write.rs

```diff
@@ -315,13 +299,90 @@ pub(crate) async fn create_writer(
     }
 }
 
+/// Serializes a single data stream in parallel and writes to an ObjectStore
+/// concurrently. Data order is preserved. In the event of an error,
+/// the ObjectStore writer is returned to the caller in addition to an error,
+/// so that the caller may handle aborting failed writes.
+async fn serialize_rb_stream_to_object_store(
+    mut data_stream: Pin<Box<dyn RecordBatchStream + Send>>,
+    mut serializer: Box<dyn BatchSerializer>,
+    mut writer: AbortableWrite<Box<dyn AsyncWrite + Send + Unpin>>,
+) -> std::result::Result<
+    (
+        Box<dyn BatchSerializer>,
+        AbortableWrite<Box<dyn AsyncWrite + Send + Unpin>>,
+        u64,
+    ),
+    (
+        AbortableWrite<Box<dyn AsyncWrite + Send + Unpin>>,
+        DataFusionError,
+    ),
+> {
+    let mut row_count = 0;
+    // Not using JoinSet here since we want to ulimately write to ObjectStore preserving file order
+    let mut serialize_tasks: Vec<JoinHandle<Result<(usize, Bytes), DataFusionError>>> =
+        Vec::new();
+    while let Some(maybe_batch) = data_stream.next().await {
+        let mut serializer_clone = match serializer.duplicate() {
+            Ok(s) => s,
+            Err(_) => {
+                return Err((
+                    writer,
+                    DataFusionError::Internal(
+                        "Unknown error writing to object store".into(),
+                    ),
+                ))
+            }
+        };
+        serialize_tasks.push(task::spawn(async move {
+            let batch = maybe_batch?;
+            let num_rows = batch.num_rows();
+            let bytes = serializer_clone.serialize(batch).await?;
+            Ok((num_rows, bytes))
+        }));
+    }
+    for serialize_result in serialize_tasks {
+        let result = serialize_result.await;
+        match result {
+            Ok(res) => {
+                let (cnt, bytes) = match res {
+                    Ok(r) => r,
+                    Err(e) => return Err((writer, e)),
+                };
+                row_count += cnt;
+                match writer.write_all(&bytes).await {
+                    Ok(_) => (),
+                    Err(_) => {
+                        return Err((
+                            writer,
+                            DataFusionError::Internal(
+                                "Unknown error writing to object store".into(),
+                            ),
+                        ))
+                    }
+                };
+            }
+            Err(_) => {
+                return Err((
+                    writer,
+                    DataFusionError::Internal(
+                        "Unknown error writing to object store".into(),
+                    ),
+                ))
+            }
+        }
+    }
+
+    Ok((serializer, writer, row_count as u64))
+}
```

Comment:
```suggestion
async fn serialize_rb_stream_to_object_store(
    mut data_stream: Pin<Box<dyn RecordBatchStream + Send>>,
    mut serializer: Box<dyn BatchSerializer>,
    mut writer: AbortableWrite<Box<dyn AsyncWrite + Send + Unpin>>,
) -> std::result::Result<
    (
        Box<dyn BatchSerializer>,
        AbortableWrite<Box<dyn AsyncWrite + Send + Unpin>>,
        u64,
    ),
    (
        AbortableWrite<Box<dyn AsyncWrite + Send + Unpin>>,
        DataFusionError,
    ),
> {
    let (tx, mut rx) =
        mpsc::channel::<JoinHandle<Result<(usize, Bytes), DataFusionError>>>(100); // buffer size of 100, adjust as needed

    let serialize_task = tokio::spawn(async move {
        while let Some(maybe_batch) = data_stream.next().await {
            match serializer.duplicate() {
                Ok(mut serializer_clone) => {
                    let handle = tokio::spawn(async move {
                        let batch = maybe_batch?;
                        let num_rows = batch.num_rows();
                        let bytes = serializer_clone.serialize(batch).await?;
                        Ok((num_rows, bytes))
                    });
                    tx.send(handle).await.map_err(|_| {
                        DataFusionError::Internal(
                            "Unknown error writing to object store".into(),
                        )
                    })?;
                    yield_now().await;
                }
                Err(_) => {
                    return Err(DataFusionError::Internal(
                        "Unknown error writing to object store".into(),
                    ))
                }
            }
        }
        Ok(serializer)
    });

    let mut row_count = 0;
    while let Some(handle) = rx.recv().await {
        match handle.await {
            Ok(Ok((cnt, bytes))) => {
                match writer.write_all(&bytes).await {
                    Ok(_) => (),
                    Err(_) => {
                        return Err((
                            writer,
                            DataFusionError::Internal(
                                "Unknown error writing to object store".into(),
                            ),
                        ))
                    }
                };
                row_count += cnt;
            }
            Ok(Err(e)) => {
                // Return the writer along with the error
                return Err((writer, e));
            }
            Err(_) => {
                // Handle task panic or cancellation
                return Err((
                    writer,
                    DataFusionError::Internal(
                        "Serialization task panicked or was cancelled".into(),
                    ),
                ));
            }
        }
    }

    let serializer = match serialize_task.await {
        Ok(Ok(serializer)) => serializer,
        Ok(Err(e)) => return Err((writer, e)),
        Err(_) => {
            return Err((
                writer,
                DataFusionError::Internal("Unknown error writing to object store".into()),
            ))
        }
    };
    Ok((serializer, writer, row_count as u64))
}
}
```
How about this?

Reply:
This is fantastic, thank you @metesynnada ! I tested this and with just a small change it resulted in identical performance with much lower memory usage due to the backpressure introduced by the channel. The FIFO tests is also passing now, so win win!

Reply:
@metesynnada Actually, I just realized that the fifo test passing is dependent on leaving the `yield_now().await` call in. That call results in the serialize tasks being run in sequence rather than parallel though, so performance is poor. I'm not quite sure how to get the fifio test to pass without the `yield_now().await` call.

Reply:
Thank you for your feedback on the code. I've taken a closer look at the use of `yield_now().await` and its implications on task execution. Let's clarify a few points:

**1- Purpose of `yield_now().await`:** The `yield_now().await` call tells the Tokio runtime, "I'm temporarily done with my current work, and you can run other tasks if they're ready." It doesn't inherently force tasks to run in sequence; instead, it provides an opportunity for the runtime to switch to another task. If no other task is ready, the current task will continue almost immediately.

**2- Parallelism vs. Concurrency:** Tokio's work-stealing scheduler allows tasks to run concurrently by default. True parallel execution requires multiple threads. If your profiling runtime is configured to use a single thread, tasks will run sequentially, even if they're spawned concurrently. The `yield_now().await` call can influence the order of task execution but doesn't force sequential execution by itself.

Based on what you've said, I would suggest checking if the profiling runtime is single-threaded. Can you also provide your configuration regarding this matter?

Reply:
I believe we are facing the same issue at https://github.com/apache/arrow-datafusion/issues/5278. We could apply the same heuristic to https://github.com/apache/arrow-datafusion/issues/6290.

Reply:
Also, is your profiling environment multi-threaded?

Reply:
Thanks for the issue links. I tried calling yield_now().await every N iterations, but the FIFO insert test only passes when N==1 (i.e. yield_now().await on every iteration). 

I just pushed up a possible option of calling yield_now() if the input is unbounded, otherwise skipping it. That allows bounded tables to benefit fully from the parallelization but still allowing fifo tests to pass as they currently do. 

The fifo test is not a multithreaded runtime, but when I benchmark performance I am running in a multithreaded runtime, yes.

Reply:
I will check the entire pull request tomorrow. Thank you for your hard work. I am also pleased that the new mechanism has contributed to the performance.